ui/test/nuxeo-document-layout.test.js

```diff
@@ -0,0 +1,232 @@
+    expect(documentLayout.element).to.be.undefined;
+  });
+
+  test('Should load a layout', async () => {
+    documentLayout = await buildLayout();
+    // check the layout was loaded
+    expect(documentLayout.element).to.not.be.null;
+    expect(documentLayout.element.tagName).to.equal('NUXEO-TEST-EDIT-LAYOUT');
+    // assert layout integrity
+    expect(documentLayout.element.document).to.equal(documentLayout.document);
+    const { children } = documentLayout.element.shadowRoot;
+    expect(children).to.have.lengthOf(3);
+    expect(children[0].tagName).to.equal('STYLE');
+    expect(children[1].tagName).to.equal('NUXEO-INPUT');
+    expect(children[2].tagName).to.equal('NUXEO-TEXTAREA');
+  });
+
+  test('Should fire a "document-layout-changed" event when the layout is loaded', async () => {
+    const ePromise = new Promise((resolve) => {
+      document.addEventListener('document-layout-changed', (e) => resolve(e));
+    });
+    documentLayout = await buildLayout();
+    // check that the event was fired from the document layout
+    const event = await ePromise;
+    expect(event.target).to.equal(documentLayout);
+    // ensure that the event bubbles and is composed
+    expect(event.bubbles).to.be.true;
+    expect(event.composed).to.be.true;
+    // verify event details
+    expect(event.detail).to.exist;
+    expect(event.detail.element).to.equal(documentLayout.element);
+    expect(event.detail.layout).to.equal(documentLayout.layout);
+  });
+
+  test('Should do input validation', async () => {
+    documentLayout = await buildLayout();
+    // check that validation fails if required fields are not filled
+    expect(documentLayout.validate()).to.be.false;
+    const myTitle = 'My Title';
+    getWidget('Title').value = myTitle;
+    // assert that validation passes if required fields are filled
+    expect(documentLayout.document.properties['dc:title']).to.equal(myTitle);
+    expect(documentLayout.validate()).to.be.true;
+  });
+
+  test('Should do custom input validation if "validate" method is available on layout', async () => {
+    documentLayout = await buildLayout(
+      buildDoc({
+        'dc:title': 'My Title',
+        'dc:description': 'My Title',
+      }),
+    );
+
+    // check that validation fails if title and description are the same (custom validation)
+    expect(typeof documentLayout.element.validate).to.equal('function');
+    sinon.spy(documentLayout.element, 'validate');
+    expect(documentLayout.validate()).to.be.false;
+    expect(documentLayout.element.validate.calledOnce).to.be.true;
+    const myDescription = 'My Description';
+    // assert that validation passes if the values differ
+    getWidget('Description').value = myDescription;
+    expect(documentLayout.document.properties['dc:description']).to.equal(myDescription);
+    expect(documentLayout.validate()).to.be.true;
+    expect(documentLayout.element.validate.calledTwice).to.be.true;
+  });
+
+  test('Should do auto-focus', async () => {
```

Comment:
`when layout contains an element with autofocus attribute` or `when layout contains an auto focusable element`? 🤷‍♂ 
It would be nice if we could keep consistency with the naming convention (this one Should **_ExpectedBehavior_** When **_StateUnderTest_** or a different one that we find better 🙂 )

Reply:
I do think we can put a when where "when the layout is stamped", but I don't think it will always make sense to have one...